Rework startup and async init logic

humphd opened the pull request:
Fixes #319, part of #60.  This can happen in 0.6.

This changes a bunch of things about how startup happens, in particular how async startup logic is handled.  The way that Remix does its initialization is different than I'm used to.  It works in 2 parts, with `server.ts` (the "server") and then everything in `app/*` (the "remix" bundle).  These produce `build/server.js` and `build/index.js` respectively.

The way that live reload happens on the server side, everything in `require.cache` gets deleted when the server restarts.  This means that module-level, in-memory variables (e.g., instances) get deleted in development (but not production).  To combat this, they recommend you cache instances on the `global` in development, and reattach them when loading again.

This is fine for instances created using synchronous code paths, but breaks down for asynchronous code.  We have a bunch of the latter, and needed a solution for setting it up once.

I've adjusted a few things to accomplish this:

1. Altered the way that `saml.server.ts` loads the IDP metadata.  We now do it from a file synchronously vs. fetching from a URL.  This turns out to be the recommended way anyway, from a security point of view.  I've had to update the `env` and Docker setup to deal with this new build arg/env requirement.  I've added the XML for dev and staging to `config/`.
2. Altered the way that `dns.server.ts` creates the hosted zone in development.  I'm not creating it dynamically on first use, inside existing `async` functions.  Only the first call will incur the extra cost of creation, and others will use the same zone id.
3. Altered the way that `logger.server.ts` was setting up the termination logging.  I only want to add these listeners once, so this manages that via a `global` in dev.

I fixed some other things along the way:

1. Bug in how the AWS URL endpoint was being set for development.  It wouldn't have worked in production
2. I've increased the default max listeners for `EventEmitter` so we don't keep get warnings.
3. I've removed the extra `await` in `saml.server.ts` for `parseLoginResponse` on `idp`
4. Improved a race condition running `npm run dev` by alter the order we call the various `dev:*` scripts vs. letting it decide on its own.

Later I'll do a bit more on this, but I don't want to change the worker queue stuff until we land @SerpentBytes's stuff in #268.

sfrunza13 reviewed (commented):
Would it be any better or safer to have the SAML metadata as a secret instead of a file under config?

SerpentBytes commented on app/lib/redis.server.ts:
```diff
@@ -5,8 +5,9 @@ import type { Redis as RedisType, RedisOptions } from 'ioredis';
-EventEmitter.defaultMaxListeners = 64;
+// fix for this warning is to increase that default. Setting this high
+// per https://github.com/OptimalBits/bull/issues/615#issuecomment-897051838
+EventEmitter.defaultMaxListeners = 10_000;
```
I am guessing as we added more Worker/Queue logic, we started getting this error more frequently. With the addition of the first worker/queue, I recall that the warning related to hitting the limits of  MaxListeners was occasional. Setting the value to max should eliminate the warning for good.

Genne23v reviewed (approved):
Looks good to me!

humphd commented:
> Would it be any better or safer to have the SAML metadata as a secret instead of a file under config?

It's not a secret, so I don't think it belongs there, no.

humphd commented:
I've rebased this in @SerpentBytes' work that just landed, and gotten rid of the last of the `init` logic in the server.  I've also moved the shutdown logic for each worker to the same file where it happens, so there isn't a module dependency between the server and remix bundle.

Due to a race condition in the startup logic with `dev`, I've stopped using the `--watch` on the server build, and instead pre-build it before we start `dev`.

SerpentBytes commented on app/queues/certificate/challenge-completer-worker.server.ts:
```diff
@@ -17,3 +17,5 @@ export const challengeCompleterWorker = new Worker<ChallengeCompleterData>(
   { connection: redis }
 );
+
+process.on('SIGINT', () => challengeCompleterWorker.close());
```
How is closing connections to workers based on signal interruption better than how we were doing things previously? By doing this, are we building towards the panic feature requested in #142.

humphd replied:
It's the same idea.  The graceful shutdown would do the same thing, but it would require us to create a connection between the server bundle and remix bundle (i.e., we'd have to get a reference to the workers in another bundle).  This way, we can do it all in the same bundle.

The SIGINT is a POSIX signal, see https://dsa.cs.tsinghua.edu.cn/oj/static/unix_signal.html#:~:text=The%20SIGINT%20signal%20is%20sent,break%22%20key%20can%20be%20used.&text=The%20SIGKILL%20signal%20is%20sent,to%20terminate%20immediately%20(kill).

SerpentBytes replied:
Thanks for clarifying.

humphd commented:
@Genne23v can you sign-off on this before I merge?

Genne23v commented:
> @Genne23v can you sign-off on this before I merge?

I just signed off
